Public/Src/Engine/Processes/SharedOpaqueOutputLogger.cs

```diff
@@ -0,0 +1,233 @@
+        ///
+        /// NOTE: this method does not validate the recorded paths in any way.  That means that each returned string may be
+        ///   - a path pointing to an absent file
+        ///   - a path pointing to a file
+        ///   - a path pointing to a directory.
+        /// 
+        /// NOTE: unless the log file was corrupted (i.e., it was produced by an instance of this class):
+        ///   - the strings in the returned enumerable are all legal paths
+        ///   - the returned collection does not contain any duplicates
+        /// </remarks>
+        public static IEnumerable<string> ReadRecordedPathsFromSharedOpaqueOutputLog(string filePath)
+        {
+            using (var fileStream = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read | FileShare.Delete))
+            using (var bxlReader = new BuildXLReader(stream: fileStream, debug: false, leaveOpen: false))
+            {
+                string nextString = null;
+                while ((nextString = ReadStringOrNull(bxlReader)) != null)
+                {
+                    yield return nextString;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Same as <see cref="ReadRecordedPathsFromSharedOpaqueOutputLog(string)"/> except that all exceptions are wrapped in <see cref="BuildXLException"/>
+        /// </summary>
+        public static string[] ReadRecordedPathsFromSharedOpaqueOutputLogWrapExceptions(string filePath)
+        {
+            try
+            {
+                return ReadRecordedPathsFromSharedOpaqueOutputLog(filePath).ToArray();
+            }
+            catch (Exception e)
+            {
+                throw new BuildXLException("Failed to read from shared opaque output log", e);
+            }
+        }
+
+        /// <summary>
+        /// <see cref="RecordFileWrite(PathTable, AbsolutePath)"/>
+        /// </summary>
+        public bool RecordFileWrite(PathTable pathTable, string absolutePath)
+            => RecordFileWrite(pathTable, AbsolutePath.Create(pathTable, absolutePath));
+
+        /// <summary>
+        /// Records that the file at location <paramref name="path"/> was written to.
+        /// 
+        /// Returns whether the path was recorded or skipped.
+        /// </summary>
+        /// <returns>
+        /// <code>true</code> if <paramref name="path"/> was recorded, <code>false</code> 
+        /// if the path was filtered out because of <see cref="RootDirectories"/>.
+        /// </returns>
+        /// <remarks>
+        /// NOT THREAD-SAFE.
+        /// </remarks>
+        public bool RecordFileWrite(PathTable pathTable, AbsolutePath path)
+        {
+            if (RootDirectories == null || RootDirectories.Any(dir => path.IsWithin(pathTable, AbsolutePath.Create(pathTable, dir))))
+            {
+                if (m_recordedPathsCache.Add(path))
+                {
+                    m_lazyBxlWriter.Value.Write(path.ToString(pathTable));

```

Comment:
>m_lazyBxlWriter.Value [](start = 20, length = 21)

Any reason why the writer is made lazy?

Reply:
Yes, the reason is that the tests can use non-existing file names if they are not going to write anything to this log.